staging/src/k8s.io/apimachinery/pkg/api/meta/interfaces.go

```diff
@@ -133,11 +135,112 @@ type RESTMapper interface {
 	ResourceSingularizer(resource string) (singular string, err error)
 }
 
+// RESTMapperWithContext allows clients to map resources to kind, and map kind and version
+// to interfaces for manipulating those objects. It is primarily intended for
+// consumers of Kubernetes compatible REST APIs as defined in docs/devel/api-conventions.md.
+//
+// The Kubernetes API provides versioned resources and object kinds which are scoped
+// to API groups. In other words, kinds and resources should not be assumed to be
+// unique across groups.
+type RESTMapperWithContext interface {
+	// KindFor takes a partial resource and returns the single match.  Returns an error if there are multiple matches
+	KindForWithContext(ctx context.Context, resource schema.GroupVersionResource) (schema.GroupVersionKind, error)
+
+	// KindsFor takes a partial resource and returns the list of potential kinds in priority order
+	KindsForWithContext(ctx context.Context, resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error)
+
+	// ResourceFor takes a partial resource and returns the single match.  Returns an error if there are multiple matches
+	ResourceForWithContext(ctx context.Context, input schema.GroupVersionResource) (schema.GroupVersionResource, error)
+
+	// ResourcesFor takes a partial resource and returns the list of potential resource in priority order
+	ResourcesForWithContext(ctx context.Context, input schema.GroupVersionResource) ([]schema.GroupVersionResource, error)
+
+	// RESTMapping identifies a preferred resource mapping for the provided group kind.
+	RESTMappingWithContext(ctx context.Context, gk schema.GroupKind, versions ...string) (*RESTMapping, error)
+	// RESTMappings returns all resource mappings for the provided group kind if no
+	// version search is provided. Otherwise identifies a preferred resource mapping for
+	// the provided version(s).
+	RESTMappingsWithContext(ctx context.Context, gk schema.GroupKind, versions ...string) ([]*RESTMapping, error)
+
+	ResourceSingularizerWithContext(ctx context.Context, resource string) (singular string, err error)
+}
+
+func RESTMapper2Context(m RESTMapper) RESTMapperWithContext {
+	if m == nil {
+		return nil
+	}
+	if m, ok := m.(RESTMapperWithContext); ok {
+		return m
+	}
+	return &restMapperWrapper{
```

Comment:
I wasn't entirely sure whether this should return a value or a pointer to satisfy the interface. Both works, so I benchmarked it. Mem allocs are the same (one alloc/call), but returning a value is 10% slower when calling one method.

What I then benchmarked is whether pointer vs value receiver makes a difference. Converting from value receiver (what I had before) to pointer receiver reduced call overhead by 6%. That's because with a value receiver, Go has to auto-generate a variant with pointer receiver and calls the value receiver through that.

That can be seen in a debugger (call stack) and when setting breakpoints:
```
(dlv) b restMapperWrapper.KindForWithContext
Command failed: Location "restMapperWrapper.KindForWithContext" ambiguous: k8s.io/apimachinery/pkg/api/meta.restMapperWrapper.KindForWithContext, k8s.io/apimachinery/pkg/api/meta.(*restMapperWrapper).KindForWithContext…
```

Conventional wisdom is to define types with value receiver because those can be called also on unmutable instances, making them more flexible.

But for types which will only ever be used via a pointer, I think pointer receiver is better for the reasons above (small performance difference, easier to debug).

I'll change to that in the entire PR.